sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs

```diff
@@ -0,0 +1,119 @@
+
+        private static readonly ISet<KnownType> TrackedTestAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute,
+            KnownType.NUnit_Framework_TestAttribute,
+            KnownType.Xunit_FactAttribute);
+
+        private static readonly ISet<KnownType> ExpectedExceptionAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_ExpectedExceptionAttribute,
+            KnownType.NUnit_Framework_ExpectedExceptionAttribute);
+
+        private static readonly ISet<KnownType> TrackedAssertionTypes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert,
+            KnownType.NUnit_Framework_Assert,
+            KnownType.Xunit_Assert);
+
+        private static readonly IEnumerable<string> TrackedAssertionStartWithMethodName = ImmutableList.Create(
+            "Assert", "Should", "Expect", "Must", "Verify");
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+                    if (methodDeclaration.Identifier.IsMissing ||
+                        methodDeclaration.Body == null ||
+                        !c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+                    var attributes = methodSymbol?.GetAttributes();
+                    if (methodSymbol == null ||
+                        !IsTestMethod(attributes) ||
+                        HasExpectedExceptionAttribute(attributes))
+                    {
+                        return;
+                    }
+
+                    var hasAnyAssertion = methodDeclaration.DescendantNodes()
+                        .OfType<InvocationExpressionSyntax>()
+                        .Any(invocation => IsAssertion(invocation, c.SemanticModel));
+                    if (!hasAnyAssertion)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation()));
+                    }
+                }, SyntaxKind.MethodDeclaration);
+        }
+
+        private static bool IsTestMethod(IEnumerable<AttributeData> attributes) =>
+            attributes.Any(attribute => attribute.AttributeClass.IsAny(TrackedTestAttributes));
+
+        private static bool HasExpectedExceptionAttribute(IEnumerable<AttributeData> attributes) =>
+            attributes.Any(attribute => attribute.AttributeClass.IsAny(ExpectedExceptionAttributes));
+
+        private static bool IsAssertion(InvocationExpressionSyntax invocation, SemanticModel model)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax' instead of 'Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)